Foundry to hardhat migration guide (#272)
NomicFoundation/hardhat-website: src/content/docs/docs/migrate-from-foundry/index.mdx

---
title: Migrate from Foundry
description: How to migrate from Foundry to Hardhat 3
sidebar:
  label: Overview
---

import { TabItem, Tabs } from "@astrojs/starlight/components";
import { Code } from "@astrojs/starlight/components";
import Install from "@hh/Install.astro";
import Run from "@hh/Run.astro";

Hardhat 3 is a full replacement for Foundry: it runs Solidity tests, understands `remappings.txt`, and can build contracts laid out the way Foundry expects.

On top of that compatibility, Hardhat 3 brings capabilities Foundry doesn't:

> - **Dependencies managed with NPM** — you have the option to keep your existing `lib/` git submodules, but with Hardhat 3 you can install dependencies as npm packages to get lockfiles, semver, and native support for transitive dependencies.

Reviewer: ```suggestion
- **Dependencies managed with NPM** — Hardhat supports both git submodules and npm dependencies. You have the option to keep your existing `lib/` git submodules, and you can also install dependencies as npm packages to get lockfiles, semver, and native support for transitive dependencies.
```

wdyt? the previous opening was a bit more negative

Author: Agreed - though I am a bit more negative git submodules!

- **Multichain testing** — run the same Solidity tests against a simulated Ethereum L1 or OP Stack chain by switching the chain type, so you catch chain-specific behavior before you deploy.

> - **The right test for the job** — write fast Solidity unit tests for contract logic, exactly as you do today, and reach for TypeScript integration tests (with viem or ethers) when you need to test against a full network.

Reviewer: ```suggestion
- **The right test for the job** — write fast Solidity unit tests for contract logic, exactly as you do today, and reach for TypeScript for integration tests (with viem or ethers) when you need to test against a full network, or to build custom scripts in a full-featured language.
```

something like that? ts is useful beyond tests

> 

Author: A different set of key selling points could be selected. They are based off of: https://hardhat.org/docs/hardhat3/whats-new

Reviewer: I think this initial list looks fine

Reviewer: I would add as the first one that you have access to a customizable stack that integrates with the whole typescript ecosystem

Author: I added this first point:

```
- **An extensible stack** — with a [plugin system](/docs/plugins/official-plugins)
  and integration with the TypeScript ecosystem, you can bring your own opinions
  to deployment, verification and other workflows.
```

This guide will take you, step by step, through migrating a Foundry project to Hardhat 3. We start with a clean Hardhat config, alongside the existing `foundry.toml`, and migrate one capability at a time.

## Before starting the migration

Before installing Hardhat 3, prepare your project.

1.  **Check Node.js version**

>     Make sure you have installed Node.js v22.10.0 or later:

Reviewer: ```suggestion
    Make sure you have installed Node.js v22.13.0 or later:
```

1.  **Install npm equivalents of your `lib/` dependencies**

    Most common dependencies are published to npm. For example:

    <Install packages="@openzeppelin/contracts" />

    For `forge-std`, install it directly from GitHub via your package manager:

    <Install packages="'github:foundry-rs/forge-std#v1.9.7'" />

2.  **Reconcile remappings**

    Hardhat 3 reads `remappings.txt` files automatically, including the ones inside npm packages and git submodules, so your existing entries keep working. Once a dependency is installed via npm, you can either:
    - Update the remapping target to point at the npm package (e.g. `@openzeppelin/contracts/=node_modules/@openzeppelin/contracts/`), or
    - Remove the remapping and update your `import` statements to use the npm path directly.

    If your project relies on `forge`-generated remappings (no `remappings.txt` is checked in), install the [`@nomicfoundation/hardhat-foundry`](/docs/reference/foundry-compatibility#the-nomicfoundationhardhat-foundry-plugin) plugin and add it to your `plugins` array.

> 3.  **Remove `lib/` once everything builds**

Reviewer: Only if everything was installed through npm

Author: Integrated.

    Re-run `hardhat build`. Once it succeeds, you can delete `lib/` and the corresponding `.gitmodules` entries.

> 

Reviewer: Something like this is missing:

If you use direct imports from the root of your project (e.g. `import "src/Foo.sol"`), please take a look at [this guide](https://hardhat.org/docs/cookbook/absolute-imports).

Author: I have added a note referring to the absolute imports guide.

## Migrating Solidity tests

Hardhat 3's Solidity Test is designed for compatibility with Foundry test suites. Solidity tests, cheatcodes, and `forge-std` are supported out of the box — no plugin required.

1.  **Confirm your test layout is picked up**

    By default, Hardhat treats every `.sol` file under `test/` as a test file, and any `.t.sol` file under `src/` (or your configured `sources` directory). A contract is treated as a test contract if it has at least one `test*` function, and `setUp()` runs before each test — the same conventions as Foundry. If your tests live somewhere else, override `paths.tests.solidity`:

    ```ts
    // hardhat.config.ts
    export default defineConfig({
      paths: {
        tests: {
          solidity: "./test",
        },
      },
    });
    ```

> 2.  **Run a single test file**

Author: I considered just running the full suite, but I think it is helpful to show the single test file capability.

Reviewer: yes, makes sense

:::

## Migrating scripts and deployments

Hardhat doesn't have a direct equivalent of `forge script`. For deployments, we recommend [Hardhat Ignition](/docs/guides/deployment), which is declarative and handles resumability, idempotency, and verification. For one-off scripts (admin actions, data migrations), define a [Hardhat task](/docs/guides/writing-tasks) and call your contracts through a network connection:

```ts
// hardhat.config.ts
import { defineConfig, task } from "hardhat/config";

const deployCounter = task("deploy-counter", "Deploys the Counter contract")
  .setInlineAction(async (_taskArgs, hre) => {
    const { viem } = await hre.network.create();
    const counter = await viem.deployContract("Counter");

    console.log("Counter deployed at:", counter.address);
  })
  .build();

export default defineConfig({
  tasks: [deployCounter],
});
```

The `viem` (or `ethers`) helpers come from a toolbox plugin; install whichever you prefer and add it to `plugins` in your config.

## Migration blockers

> If your migration is blocked by a missing feature, plugin, or cheatcode, please let us know [in this issue](https://github.com/NomicFoundation/hardhat/issues/7207).

Author: Should we open a separate issue for Foundry migration issues?

Reviewer: i think so

Author: I have opened this issue: https://github.com/NomicFoundation/hardhat/issues/8347, and referred to it instead.
